### PR TITLE
New option for file level overwrites

### PR DESCRIPTION
**Pull request by dennis-behm**
As described in #83 , we would like to provide a new option to allow file level overwrites through a dedicated properties file.

This PR has the following components:
- The application can manage a dedicated property file in a subfolder for each build file. The property file can contain any build property which is retrieved in the language scripts via the `getFileProperty()` Method. `application.properties` defines the location of the property files.
- The build team can activate this feature in build-conf. Additionally, property overwrites are checked against a whitelist of properties. See the readme of build-conf.

Attention: Use this feature with care - if you go with the approach of 1 property file for each source file (and for all source file you have), and one day you have to modify a property, will require to modify hundreds or thousands property files. We recommend to use it for exceptions.

### Sample usage:

For the **application team**: Within the MortgageApplication application, a new subfolder contains property files on a file basis:
image

The file name has to match with the file name of the buildfile.

sample contents of `MortgageApplication/properties/epsnbrvl.properties`
```
cobol_compileParms=HUGO
cobol_compilerVersion=V4
```

Additionally, the location of the properties file need to be specified in the `application.properties` in application-conf
```
#
# location for file level overwrites relative to {props.application}
# {props.application}/{propertiesFileLocation}
propertiesFileLocation = properties

```

The **build admins** activates the feature in build-conf as well as defines the whitelist of properties which are allowed to be overwritten.
```
# boolean condition to activate looking for file level property files 
allowFileLevelOverwrites=true

# Comma-separated list of permitted file level properties, which are retrieved via props.getFileProperty api
whitelistFileLevelOverwrites=cobol_compileParms
```

Sample output:

```
MortgageApplication/cobol/epsnbrvl.cbl
** Invoking build scripts according to build order: BMS.groovy,Cobol.groovy,LinkEdit.groovy
** Building files mapped to Cobol.groovy script
required props = cobol_srcPDS,cobol_cpyPDS,cobol_objPDS,cobol_loadPDS,cobol_compiler,cobol_linkEditor,cobol_tempOptions,applicationOutputsCollectionName,  SDFHCOB,SDFHLOAD,SDSNLOAD,SCEELKED
** Creating / verifying build dataset DBEHM.DBB.BUILD.COBOL
** Creating / verifying build dataset DBEHM.DBB.BUILD.COPY
** Creating / verifying build dataset DBEHM.DBB.BUILD.OBJ
** Creating / verifying build dataset DBEHM.DBB.BUILD.DBRM
** Creating / verifying build dataset DBEHM.DBB.BUILD.LOAD
*** Building file MortgageApplication/cobol/epsnbrvl.cbl
>>> ** Loaded file level property overwrites for MortgageApplication/cobol/epsnbrvl.cbl:
>>>>   [cobol_compileParms:HUGO, cobol_compilerVersion:V4]
*** Creating dependency resolver for MortgageApplication/cobol/epsnbrvl.cbl with [{"library": "SYSLIB", "searchPath": [ {"sourceDir": "/u/dbehm/test/dbb-zappbuild/samples", "directory": "MortgageApplication/copybook"} ]                }] rules
*** Scanning file with the default scanner
*** Resolution rules for MortgageApplication/cobol/epsnbrvl.cbl:
{"library":"SYSLIB","searchPath":[{"sourceDir":"\/u\/dbehm\/test\/dbb-zappbuild\/samples","directory":"MortgageApplication\/copybook"}]}
*** Physical dependencies for MortgageApplication/cobol/epsnbrvl.cbl:
{"sourceDir":"\/u\/dbehm\/test\/dbb-zappbuild\/samples","lname":"EPSNBRPM","library":"SYSLIB","file":"MortgageApplication\/copybook\/epsnbrpm.cpy","category":"COPY","resolved":true}
>>> ** File level overwrite found for MortgageApplication/cobol/epsnbrvl.cbl - cobol_compileParms: HUGO
Cobol compiler parms for MortgageApplication/cobol/epsnbrvl.cbl = HUGO
>>> ** File level overwrite found for MortgageApplication/cobol/epsnbrvl.cbl - cobol_compilerVersion: V4
>>> *! Overwrite for cobol_compilerVersion not allowed. Not on whitelist. See build property whitelistFileLevelOverwrites.
** Writing build report data to /u/dbehm/test/work/Mortgage/BuildReport.json
```

### CODE REVIEW

File: build-conf/README.md [hunk 1]
```diff
@@ -44,7 +44,8 @@ dbb.RepositoryClient.url | DBB configuration property for web application URL.
 dbb.RepositoryClient.userId | DBB configuration property for web application logon id.  ***Can be overridden by build.groovy option -id, --id***
 dbb.RepositoryClient.password | DBB configuration property for web application logon password.  ***Can be overridden by build.groovy option -pw, --pw***
 dbb.RepositoryClient.passwordFile | DBB configuration property for web application logon password file.  ***Can be overridden by build.groovy option -pf, --pf***
-
+allowFileLevelOverwrites| Configuration property to enable zAppBuild to search for property files for each buildable file defined by the propertiesFileLocation property in application-conf 
+whitelistFileLevelOverwrites | Comma-separated list of file level properties, which can be overridden by a file level property file retrieved via props.getFileProperty api in the lanaguage scripts
```
Comment (jbyibm) [inline build-conf/README.md:48]: typo -> lanaguage -> language

File: build-conf/build.properties [hunk 2]
```diff
@@ -109,6 +109,9 @@ dbb.gateway.regionSize=
 # 16 - Log information to the system console
 dbb.gateway.logLevel=2
 
+# boolean condition (default:false) to activate looking for file level property files
+allowFileLevelOverwrites=false
 
-
+# Manage a comma-separated list of allowed file level properties, which are retrieved via props.getFileProperty api
+whitelistFileLevelOverwrites=
```
Comment (jbyibm) [inline build-conf/build.properties:116]: What is the default ? All ?

File: utilities/BuildUtilities.groovy [hunk 3]
```diff
@@ -370,3 +371,98 @@ def createDatasets(String[] datasets, String options) {
...
+				if (props.verbose) println("** File level overwrite found for $buildFile - $prop: $fileLevelProperty")
+				if (props.verbose)
+					return fileLevelProperty
+			}
+		}
+	}
+
+	//return DBB File Property or Global property
+	return props.getFileProperty(prop, buildFile)
+}
```
Comment (jbyibm) [inline utilities/BuildUtilities.groovy]: This method is double.

File: utilities/BuildUtilities.groovy [hunk 4]
```diff
@@ -370,3 +371,98 @@ def createDatasets(String[] datasets, String options) {
...
+
+/*
+ * groovy metaprogramming to return properties to language scripts
+ *  - verifies if a propertiesfile for the buildFile exists and overwrites settings
+ */
+static def propertyMissing(String name) {
+	return props.getProperty(name)
+}
+
```
Comment (jbyibm) [inline utilities/BuildUtilities.groovy:409]: You introduce a behavior here that can complexity the debugging. This simple test case raise no error for balibalo.

```
@groovy.transform.BaseScript com.ibm.dbb.groovy.ScriptLoader baseScript
import com.ibm.dbb.build.*
import groovy.transform.*

// define script properties
@Field static BuildProperties props = BuildProperties.getInstance()

balibalo

/*
 * groovy metaprogramming to return properties to language scripts
 *  - verifies if a propertiesfile for the buildFile exists and overwrites settings
 */
static def propertyMissing(String name) {
        return props.getProperty(name)
}

```
```
>$DBB_HOME/bin/groovyz test.groovy
** Build finished
```

You can add an assert if `props.getProperty(name)`is null.

File: utilities/BuildUtilities.groovy [hunk 5]
```diff
@@ -370,3 +371,98 @@ def createDatasets(String[] datasets, String options) {
...
+/*
+ * get filename without file extension
+ */
+def getFileNameWithoutExtension(String buildFile){
+	int pos = buildFile.lastIndexOf(".");
+	int posFolder = buildFile.lastIndexOf("/")
+
+	if (pos > 0) {
+		return buildFile.substring(posFolder+1, pos);
+	}
```
Comment (jbyibm) [inline utilities/BuildUtilities.groovy]: You are already working on the handle of the extension for multiple source code (COBOL, DBD, PSD, ..) with same file name.

**Review by jbyibm — CHANGES_REQUESTED** [review]

**Review by jbyibm — APPROVED** [review]

**Review by jbyibm — COMMENTED** [review]
Look my comments

File: utilities/BuildUtilities.groovy (hunk 4 above)
Comment (dennis-behm) [reply utilities/BuildUtilities.groovy:409]: Hi @jbyibm, you are right.

Already today, there are several fields which are used and are tested against NULL.

Additionally, `out` seems to be an object being returned when creating a new instance of an object ... A bit odd ... 
 

```
static def propertyMissing(String name) {
	def optionalParms = ["out", "errPrefix", "debug", "runzTests", "team", "logEncoding", "RESLIB", "userBuild"]
	if (!optionalParms.contains(name)) assert props."$name" : "*! Missing property / field '$name'"
	return props.getProperty(name)
}
```

**Comment by dennis-behm** [comment]
Convert the pull request to be a **DRAFT**.

I am facing an issue with the Groovy metaprogramming.
The particular case is ... if a build fails the status is set via

```
				props.error = "true"
```

But due to the metaprogramming, we need to make sure to update the DBB BuildProperties table ....

**Comment by dennis-behm** [comment]
I am closing the PR, while it requires a different way to implement this. Perusing the Groovy Metaprogramming would increase complexity of the build framework. 

Notes for the new implementation. Leverage the DBB API's to add file patterns programmatically through the `BuildProperties` API

https://www.ibm.com/docs/api/v1/content/SS6T76_1.1.0/javadoc/com/ibm/dbb/build/BuildProperties.html#addFilePatterns-java.lang.String-java.lang.String-java.util.List-

**Pull request closed without merging**